toolchain/check/eval.cpp

```diff
@@ -4,31 +4,82 @@
 
 #include "toolchain/check/eval.h"
 
+#include "toolchain/sem_ir/ids.h"
 #include "toolchain/sem_ir/typed_insts.h"
 #include "toolchain/sem_ir/value_stores.h"
 
 namespace Carbon::Check {
 
+namespace {
+// The evaluation phase for an expression, computed by evaluation. These are
+// ordered so that the phase of an expression is the numerically highest phase
+// of its constituent evaluations.
+enum class Phase : uint8_t {
+  // Value could be entirely and concretely computed.
+  Template,
+  // Evaluation encountered a reference to a symbolic binding.
+  Symbolic,
+  // Evaluation encountered an error, which is treated as being potentially
+  // constant, but with an unknown phase.
+  Error,
+  // Evaluation encountered a non-constant construct.
+  Runtime,
+};
+}  // namespace
+
+// Gets the phase in which the value of a constant will become available.
+static auto GetPhase(SemIR::ConstantId constant_id) -> Phase {
+  if (!constant_id.is_constant()) {
+    return Phase::Runtime;
+  } else if (constant_id.is_template()) {
+    if (constant_id.inst_id() == SemIR::InstId::BuiltinError) {
+      return Phase::Error;
+    }
+    return Phase::Template;
+  } else {
+    return Phase::Symbolic;
+  }
+}
+
+// Returns the later of two phases.
+static auto LatestPhase(Phase a, Phase b) -> Phase {
+  return Phase(std::max(static_cast<uint8_t>(a), static_cast<uint8_t>(b)));
+}
+
+// Forms a `constant_id` describing a given evaluation result.
+static auto MakeConstantResult(Context& context, SemIR::Inst inst, Phase phase)
+    -> SemIR::ConstantId {
+  switch (phase) {
+    case Phase::Template:
+    case Phase::Symbolic:
+      return context.AddConstant(inst, phase == Phase::Symbolic);
```

Comment:
This comparison feels a little odd with the switch, but maybe that's just me.

Reply:
I'm indifferent. Switched :-)